File: c++/Minimum Number of Operations to Move All Balls to Each Box.cpp
#include <iostream>
#include <string>
#include <vector>

class Solution {
    public:
    std::vector<int> minOperations(const std::string& boxes) {
        int n = boxes.size();
        std::vector<int> result(n, 0);

        int moves = 0, count = 0;
        for (int i = 0; i < n; ++i) {
            result[i] += moves;
            count += boxes[i] == '1' ? 1 : 0;
            moves += count;
        }

        moves = 0, count = 0;
        for (int i = n - 1; i >= 0; --i) {
            result[i] += moves;
            count += boxes[i] == '1' ? 1 : 0;
            moves += count;
        }

        return result;
    }
};
// Test cases
int main() {
    Solution solution;
    // case 1
    std::string boxes1 = "110";
    std::vector<int> result1 = solution.minOperations(boxes1);
    std::cout << "Out put of test case 1:" << std::endl; 
    for(int num : result1) std::cout << num << " "; // Expected Out Put: 1,1,3 
    std::cout << std::endl;
    // case 2
    std::string boxes2 = "001011";
    std::vector<int> result2 = solution.minOperations(boxes2);
    std::cout << "Out put of test case 1:" << std::endl;
    for(int num : result2) std::cout << num << " "; // Expected Out Put: 11,8,5,4,3,4
    std::cout << std::endl;

    return 0;
}

File: python/Minimum Number of Operations to Move All Balls to Each Box.py
from typing import List

class Solution:
    def minOperations(self, boxes: str) -> List[int]:
        n = len(boxes)
        result = [0] * n

        moves, count = 0, 0
        for i in range(n):
            result[i] += moves
            count += 1 if boxes[i] == '1' else 0
            moves += count

        moves, count = 0, 0
        for i in range(n - 1, -1, -1):
            result[i] += moves
            count += 1 if boxes[i] == '1' else 0
            moves += count

        return result
# Test cases
if __name__ == "__main__":
    solution = Solution()
if __name__ == "__main__":
    solution = Solution()
    # Case 1
    boxes1 = "110"
    result1 = solution.minOperations(boxes1)
    print("Output of Test Case 1:")
    print(f"Input: {boxes1}")
    print(f"Output: {' '.join(map(str, result1))}")  # Expected Output: 1 1 3
    # Case 2
    boxes2 = "001011"
    result2 = solution.minOperations(boxes2)
    print("\nOutput of Test Case 2:")
    print(f"Input: {boxes2}")
    print(f"Output: {' '.join(map(str, result2))}")  # Expected Output: 11 8 5 4 3
